File: src/csharp/Intel/Iced/Intel/BlockEncoderInternal/XbeginInstr.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if ENCODER && BLOCK_ENCODER
using System;
using System.Diagnostics;

namespace Iced.Intel.BlockEncoderInternal {
	/// <summary>
	/// Xbegin instruction
	/// </summary>
	sealed class XbeginInstr : Instr {
		Instruction instruction;
		TargetInstr targetInstr;
		InstrKind instrKind;
		readonly byte shortInstructionSize;
		readonly byte nearInstructionSize;

		enum InstrKind : byte {
			Unchanged,
			Rel16,
			Rel32,
			Uninitialized,
		}

		public XbeginInstr(BlockEncoder blockEncoder, Block block, in Instruction instruction)
			: base(block, instruction.IP) {
			this.instruction = instruction;
			instrKind = InstrKind.Uninitialized;

			Instruction instrCopy;

			if (!blockEncoder.FixBranches) {
				instrKind = InstrKind.Unchanged;
				instrCopy = instruction;
				instrCopy.NearBranch64 = 0;
				Size = blockEncoder.GetInstructionSize(instrCopy, 0);
			}
			else {
				instrCopy = instruction;
				instrCopy.InternalSetCodeNoCheck(Code.Xbegin_rel16);
				instrCopy.NearBranch64 = 0;
				shortInstructionSize = (byte)blockEncoder.GetInstructionSize(instrCopy, 0);

				instrCopy = instruction;
				instrCopy.InternalSetCodeNoCheck(Code.Xbegin_rel32);
				instrCopy.NearBranch64 = 0;
				nearInstructionSize = (byte)blockEncoder.GetInstructionSize(instrCopy, 0);

				Size = nearInstructionSize;
			}
		}

		public override void Initialize(BlockEncoder blockEncoder) =>
			targetInstr = blockEncoder.GetTarget(instruction.NearBranchTarget);

		public override bool Optimize(ulong gained) => TryOptimize(gained);

		bool TryOptimize(ulong gained) {
			if (instrKind == InstrKind.Unchanged || instrKind == InstrKind.Rel16) {
				Done = true;
				return false;
			}

			var targetAddress = targetInstr.GetAddress();
			var nextRip = IP + shortInstructionSize;
			long diff = (long)(targetAddress - nextRip);
			diff = CorrectDiff(targetInstr.IsInBlock(Block), diff, gained);
			if (short.MinValue <= diff && diff <= short.MaxValue) {
				instrKind = InstrKind.Rel16;
				Size = shortInstructionSize;
				return true;
			}

			instrKind = InstrKind.Rel32;
			Size = nearInstructionSize;
			return false;
		}

		public override string? TryEncode(Encoder encoder, out ConstantOffsets constantOffsets, out bool isOriginalInstruction) {
			switch (instrKind) {
			case InstrKind.Unchanged:
			case InstrKind.Rel16:
			case InstrKind.Rel32:
				isOriginalInstruction = true;
				if (instrKind == InstrKind.Unchanged) {
					// nothing
				}
				else if (instrKind == InstrKind.Rel16)
					instruction.InternalSetCodeNoCheck(Code.Xbegin_rel16);
				else {
					Debug.Assert(instrKind == InstrKind.Rel32);
					instruction.InternalSetCodeNoCheck(Code.Xbegin_rel32);
				}
				instruction.NearBranch64 = targetInstr.GetAddress();
				if (!encoder.TryEncode(instruction, IP, out _, out var errorMessage)) {
					constantOffsets = default;
					return CreateErrorMessage(errorMessage, instruction);
				}
				constantOffsets = encoder.GetConstantOffsets();
				return null;

			case InstrKind.Uninitialized:
			default:
				throw new InvalidOperationException();
			}
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/enc/XbeginInstr.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.enc;

import com.github.icedland.iced.x86.Code;
import com.github.icedland.iced.x86.Instruction;

final class XbeginInstr extends Instr {
	private Instruction instruction;
	private TargetInstr targetInstr;
	private byte instrKind;
	private final byte shortInstructionSize;
	private final byte nearInstructionSize;

	private static final class InstrKind {
		static final byte UNCHANGED = 0;
		static final byte REL16 = 1;
		static final byte REL32 = 2;
		static final byte UNINITIALIZED = 3;
	}

	XbeginInstr(BlockEncoder blockEncoder, Block block, Instruction instruction) {
		super(block, instruction.getIP());
		this.instruction = instruction;
		instrKind = InstrKind.UNINITIALIZED;

		Instruction instrCopy;

		if (!blockEncoder.fixBranches()) {
			instrKind = InstrKind.UNCHANGED;
			instrCopy = instruction.copy();
			instrCopy.setNearBranch64(0);
			size = blockEncoder.getInstructionSize(instrCopy, 0);
			shortInstructionSize = 0;
			nearInstructionSize = 0;
		}
		else {
			instrCopy = instruction.copy();
			instrCopy.setCode(Code.XBEGIN_REL16);
			instrCopy.setNearBranch64(0);
			shortInstructionSize = (byte)blockEncoder.getInstructionSize(instrCopy, 0);

			instrCopy.setCode(Code.XBEGIN_REL32);
			instrCopy.setNearBranch64(0);
			nearInstructionSize = (byte)blockEncoder.getInstructionSize(instrCopy, 0);

			size = nearInstructionSize;
		}
	}

	@Override
	void initialize(BlockEncoder blockEncoder) {
		targetInstr = blockEncoder.getTarget(instruction.getNearBranchTarget());
	}

	@Override
	boolean optimize(long gained) {
		return tryOptimize(gained);
	}

	private boolean tryOptimize(long gained) {
		if (instrKind == InstrKind.UNCHANGED || instrKind == InstrKind.REL16) {
			done = true;
			return false;
		}

		long targetAddress = targetInstr.getAddress();
		long nextRip = ip + shortInstructionSize;
		long diff = targetAddress - nextRip;
		diff = correctDiff(targetInstr.isInBlock(block), diff, gained);
		if (-0x8000 <= diff && diff <= 0x7FFF) {
			instrKind = InstrKind.REL16;
			size = shortInstructionSize;
			return true;
		}

		instrKind = InstrKind.REL32;
		size = nearInstructionSize;
		return false;
	}

	@Override
	String tryEncode(Encoder encoder, TryEncodeResult result) {
		switch (instrKind) {
		case InstrKind.UNCHANGED:
		case InstrKind.REL16:
		case InstrKind.REL32:
			result.isOriginalInstruction = true;
			Instruction instruction = this.instruction.copy();
			if (instrKind == InstrKind.UNCHANGED) {
				// nothing
			}
			else if (instrKind == InstrKind.REL16)
				instruction.setCode(Code.XBEGIN_REL16);
			else {
				assert instrKind == InstrKind.REL32 : instrKind;
				instruction.setCode(Code.XBEGIN_REL32);
			}
			instruction.setNearBranch64(targetInstr.getAddress());
			Object encResult = encoder.tryEncode(instruction, ip);
			if (encResult instanceof String) {
				return createErrorMessage((String)encResult, instruction);
			}
			result.constantOffsets = encoder.getConstantOffsets();
			return null;

		case InstrKind.UNINITIALIZED:
		default:
			throw new UnsupportedOperationException();
		}
	}
}
